### PR TITLE
 server: fix insert zero timestamp bug with prepared statement

### LINKED ISSUE #7501: insert timestamp failure with 0000-00-00 00:00:00.000000 in prepare statement
1. What did you do?

```
    CREATE TABLE timestamp_insert ( id INT NOT NULL PRIMARY KEY, tstamp TIMESTAMP NOT NULL )
```
then prepare and exec a insert 0000-00-00 00:00:00.000000 into `tstamp`

```
    stmt= const_cast<char *>("INSERT INTO timestamp_insert ( id, tstamp ) VALUES ( ?, ? )");
    if(mysql_stmt_prepare(mysql_stmt, stmt, strlen(stmt)))
    {
        my_chk_stmt(&mysql, mysql_stmt, __LINE__);
    }

    memset(mysql_bind_param, 0, sizeof(mysql_bind_param));
    mysql_bind_param[0].buffer_type= MYSQL_TYPE_LONG;
    mysql_bind_param[0].buffer= (char*)&id;
    mysql_bind_param[0].is_null= 0;
    mysql_bind_param[0].length= 0;
    mysql_bind_param[1].buffer_type= MYSQL_TYPE_TIMESTAMP;
    mysql_bind_param[1].buffer= &tstamp;
    mysql_bind_param[1].is_null= 0;
    mysql_bind_param[1].length= 0;
    if(mysql_stmt_bind_param(mysql_stmt, mysql_bind_param))
    {
        my_chk_stmt(&mysql, mysql_stmt, __LINE__);
    }

    id= 2;
    tstamp.year= 0;
    tstamp.month= 0;
    tstamp.day= 0;
    tstamp.hour= 0;
    tstamp.minute= 0;
    tstamp.second= 0;
    tstamp.second_part= 0;

    if(mysql_stmt_execute(mysql_stmt))
    {
        my_chk_stmt(&mysql, mysql_stmt, __LINE__);
    }
```

2. What did you expect to see?

```
success
```

3. What did you see instead?

```
1292 : invalid time format: '0'
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
2.1.0.rc1
```

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

fixes #7501

### What is changed and how it works?

in protocol layer return `"types.ZeroDatetimeStr"` instead of `"0"` to make later `types.parseDatetime` happy.

### Check List 

Tests 

 - Unit test
 - Manual test 

```
    mysql_stmt= mysql_stmt_init(&mysql);
    if(!mysql_stmt)
    {
        my_chk_stmt(&mysql, mysql_stmt, __LINE__);
    }

    stmt= const_cast<char *>("INSERT INTO timestamp_insert ( id, tstamp ) VALUES ( ?, ? )");
    if(mysql_stmt_prepare(mysql_stmt, stmt, strlen(stmt)))
    {
        my_chk_stmt(&mysql, mysql_stmt, __LINE__);
    }

    memset(mysql_bind_param, 0, sizeof(mysql_bind_param));
    mysql_bind_param[0].buffer_type= MYSQL_TYPE_LONG;
    mysql_bind_param[0].buffer= (char*)&id;
    mysql_bind_param[0].is_null= 0;
    mysql_bind_param[0].length= 0;
    mysql_bind_param[1].buffer_type= MYSQL_TYPE_TIMESTAMP;
    mysql_bind_param[1].buffer= &tstamp;
    mysql_bind_param[1].is_null= 0;
    mysql_bind_param[1].length= 0;
    if(mysql_stmt_bind_param(mysql_stmt, mysql_bind_param))
    {
        my_chk_stmt(&mysql, mysql_stmt, __LINE__);
    }

    tstamp.year= 0;
    tstamp.month= 0;
    tstamp.day= 0;
    tstamp.hour= 0;
    tstamp.minute= 0;
    tstamp.second= 0;
    tstamp.second_part= 0;

    if(mysql_stmt_execute(mysql_stmt))
    {
        my_chk_stmt(&mysql, mysql_stmt, __LINE__);
    }
```

Code changes

 - impl change

Side effects

 - no

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7506)

### CODE REVIEW

File: server/conn_stmt.go [hunk 1]
```diff
@@ -354,7 +354,7 @@ func parseStmtArgs(args []interface{}, boundParams [][]byte, nullBitmap, paramTy
 			pos++
 			switch length {
 			case 0:
-				args[i] = "0"
+				args[i] = "00000000000000"
```
Comment (shenli) [inline server/conn_stmt.go]: Use a constant.

File: server/conn_stmt_test.go [hunk 2]
```diff
@@ -120,7 +120,7 @@ func (ts ConnTestSuite) TestParseStmtArgs(c *C) {
 				[]byte{0x00},
 			},
 			nil,
-			"0",
+			"00000000000000",
```
Comment (shenli) [inline server/conn_stmt_test.go]: ditto

**Comment by jackysp** [comment]
Why not add a case for `types.parseDatetime`?

**Comment by lysu** [comment]
@jackysp  we don't modify that, I think modified ut cover first half part, and add a second half part integration test using `'prepare s from '....''` to re-display and show fix result is enough

**Pull request merged**